proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2739,19 +2739,20 @@ lemma epQueue_tail_sign[simp]:
 (* Clag from cancelSignal_ccorres_helper *)
 
 lemma cancelIPC_ccorres_helper:
-  "ccorres dc xfdc (invs' and (\<lambda>s. sym_refs (state_refs_of' s)) and
+  "\<lbrakk>epptr = Ptr ep\<rbrakk> \<Longrightarrow>
+   ccorres dc xfdc (invs' and (\<lambda>s. sym_refs (state_refs_of' s)) and
          st_tcb_at' (\<lambda>st. (isBlockedOnSend st \<or> isBlockedOnReceive st)
                             \<and> blockingObject st = ep) thread
         and ko_at' ep' ep)
-        {s. epptr_' s = Ptr ep}
```

Comment:
Moving this from the preconditions to the assumptions might seem convenient here, but in general this move is against the flow of automation, e.g. if you want to use `ctac`; the idea is preconditions bubble up the stack and are resolved at the end. Is there any reason you had to change this?

Reply:
I made this change because I had a feeling that I should prefer the more structured-seeming `csymbr`  over `ccorres_symb_exec_r`, and using a value rather than a precondition on a variable in this helper lemma made it easier to use `csymbr` around it in the main proof. I've changed this lemma back to using the precondition, and I see how that results in the composition of the helper lemma and the main proof fit better into the ccorres proof flow.

Reply:
Cool. You've now managed the C side of this, but you've introduced a similar thing on the Haskell side in `cancelIPC_ccorres_helper` via `epptr = Ptr ep`. It might make sense to do the same manoeuvre there. Traditionally that's been done by adding a `and K (epptr = Ptr ep)` to the Haskell precondition side, and then if you really need that fact immediately in the proof, using `rule ccorres_grab_asm` as the first proof step.